packages/libs/src/sdk/services/Solana/programs/PaymentRouterClient/PaymentRouterClient.ts

```diff
@@ -0,0 +1,251 @@
+      paymentRouterPdaBump: this.programAccountBumpSeed,
+      recipients,
+      amounts,
+      totalAmount,
+      programId: this.programId
+    })
+  }
+
+  public async createMemoInstruction(params: CreateMemoInstructionRequest) {
+    const {
+      contentType,
+      contentId,
+      blockNumber,
+      buyerUserId,
+      accessType,
+      signer
+    } = await parseParams(
+      'createMemoInstructionSchema',
+      CreateMemoInstructionSchema
+    )(params)
+    const memoString = `${contentType}:${contentId}:${blockNumber}:${buyerUserId}:${accessType}`
+    return new TransactionInstruction({
+      keys: signer
+        ? [{ pubkey: signer, isSigner: true, isWritable: true }]
+        : [],
+      programId: new PublicKey('Memo1UhkJRfHyvLMcVucJwxXeuD728EqVDDwQDxFMNo'),
+      data: Buffer.from(memoString)
+    })
+  }
+
+  public async createPurchaseContentInstructions(
+    params: CreatePurchaseContentInstructionsRequest
+  ) {
+    const {
+      amount,
+      mint,
+      splits,
+      total,
+      contentId,
+      contentType,
+      blockNumber,
+      buyerUserId,
+      accessType
+    } = await parseParams(
+      'createPurchaseContentInstructions',
+      CreatePurchaseContentInstructionsSchema
+    )(params)
+    return [
+      this.createTransferInstruction({ amount, mint }),
+      this.createRouteInstruction({ splits, total, mint }),
+      this.createMemoInstruction({
+        contentId,
+        contentType,
+        blockNumber,
+        buyerUserId,
+        accessType
+      })
+    ]
+  }
+
+  /**
+   * Creates or gets the intermediate funds token account for the program.
+   * Only needs to be created once per mint.
+   * @see {@link https://github.com/solana-labs/solana-program-library/blob/d72289c79a04411c69a8bf1054f7156b6196f9b3/token/js/src/actions/getOrCreateAssociatedTokenAccount.ts getOrCreateAssociatedTokenAccount}
+   */
+  public async getOrCreateProgramTokenAccount(
+    params: GetOrCreateProgramTokenAccountRequest
+  ): Promise<Account> {
+    const args = await parseParams(
+      'getOrCreateProgramTokenAccount',
+      GetOrCreateProgramTokenAccountSchema
+    )(params)
+
+    // Check for cached account
+    const existingTokenAccount = this.existingTokenAccounts[args.mint]
+    if (existingTokenAccount) {
+      return existingTokenAccount
+    }
+
+    const mint = this.mints[args.mint]
+    if (!mint) {
+      throw new Error(`Mint ${args.mint} not configured`)
+    }
+    const associatedTokenAdddress = getAssociatedTokenAddressSync(
+      mint,
+      this.programAccount,
+      true
+    )
+
+    let account: Account | null = null
+    try {
+      account = await getAccount(this.connection, associatedTokenAdddress)
+      this.existingTokenAccounts[args.mint] = account
+    } catch (error: unknown) {
+      if (error instanceof TokenAccountNotFoundError) {
+        // As this isn't atomic, it's possible others can create associated accounts meanwhile.
+        try {
```

Comment:
should we consider retries on this?

Reply:
I don't think so. retries will be handled by relay